File: sorting/programmers_failure_rate.java
/*
테스트 1 〉	통과 (0.51ms, 74.8MB)
테스트 2 〉	통과 (0.86ms, 78.2MB)
테스트 3 〉	통과 (1.77ms, 76.8MB)
테스트 4 〉	통과 (3.05ms, 99.4MB)
테스트 5 〉	통과 (4.31ms, 97.8MB)
테스트 6 〉	통과 (0.76ms, 72.7MB)
테스트 7 〉	통과 (1.32ms, 77MB)
테스트 8 〉	통과 (3.10ms, 91.2MB)
테스트 9 〉	통과 (4.41ms, 90.9MB)
테스트 10 〉	통과 (2.73ms, 82.7MB)
테스트 11 〉	통과 (3.23ms, 86.8MB)
테스트 12 〉	통과 (3.26ms, 90.2MB)
테스트 13 〉	통과 (3.77ms, 86.6MB)
테스트 14 〉	통과 (0.61ms, 78.4MB)
테스트 15 〉	통과 (1.72ms, 78.4MB)
테스트 16 〉	통과 (1.16ms, 81.4MB)
테스트 17 〉	통과 (1.82ms, 84.2MB)
테스트 18 〉	통과 (1.18ms, 90.2MB)
테스트 19 〉	통과 (0.66ms, 76.7MB)
테스트 20 〉	통과 (1.43ms, 80.9MB)
테스트 21 〉	통과 (2.42ms, 87.9MB)
테스트 22 〉	통과 (4.00ms, 103MB)
테스트 23 〉	통과 (3.50ms, 83.1MB)
테스트 24 〉	통과 (3.34ms, 93.5MB)
테스트 25 〉	통과 (0.52ms, 66MB)
테스트 26 〉	통과 (0.56ms, 77.1MB)
테스트 27 〉	통과 (0.53ms, 76.5MB)
*/

import java.util.*;

// 실패율이 높은 스테이지부터 내림차순으로

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] stcnts = new int[N+2]; // N+1;
        int[] answer = new int[N];
        
        for (int st : stages) 
            stcnts[st]++;
        
        List<Stage> list = new ArrayList<Stage>();
        
        int npeople = stages.length;
        for (int i = 1; i <= N; i++) {
            if (npeople == 0)
                list.add(new Stage(i, (float)0));
            else
                list.add(new Stage(i, (float)stcnts[i] / npeople));
            npeople -= stcnts[i];
        }
        
        Collections.sort(list, new Comparator<Stage>(){
            @Override
            public int compare(Stage s1, Stage s2) {
                if (s2.failrate.compareTo(s1.failrate) == 0) {
                    return s1.idx.compareTo(s2.idx);
                }
                return s2.failrate.compareTo(s1.failrate);
            }
        });
        
        
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i).idx;
        }
        
        return answer;
    }
    
    static class Stage{
        Integer idx;
        Float failrate;
        public Stage(Integer idx, Float failrate) {
            this.idx = idx;
            this.failrate = failrate;
        }
    }
}

File: sorting/programmers_failure_rate.cpp
/*
테스트 1 〉	통과 (0.02ms, 3.72MB)
테스트 2 〉	통과 (0.02ms, 4.14MB)
테스트 3 〉	통과 (0.25ms, 4.18MB)
테스트 4 〉	통과 (1.06ms, 6.5MB)
테스트 5 〉	통과 (2.08ms, 9.92MB)
테스트 6 〉	통과 (0.05ms, 4.14MB)
테스트 7 〉	통과 (0.13ms, 4.11MB)
테스트 8 〉	통과 (1.05ms, 6.52MB)
테스트 9 〉	통과 (2.12ms, 9.87MB)
테스트 10 〉	통과 (1.11ms, 6.26MB)
테스트 11 〉	통과 (1.37ms, 6.58MB)
테스트 12 〉	통과 (1.52ms, 7.73MB)
테스트 13 〉	통과 (1.67ms, 8.35MB)
테스트 14 〉	통과 (0.02ms, 3.66MB)
테스트 15 〉	통과 (0.70ms, 5.29MB)
테스트 16 〉	통과 (0.35ms, 4.37MB)
테스트 17 〉	통과 (0.68ms, 5.11MB)
테스트 18 〉	통과 (0.34ms, 4.39MB)
테스트 19 〉	통과 (0.08ms, 3.86MB)
테스트 20 〉	통과 (0.49ms, 4.79MB)
테스트 21 〉	통과 (0.94ms, 6.16MB)
테스트 22 〉	통과 (2.69ms, 9.86MB)
테스트 23 〉	통과 (3.01ms, 9.73MB)
테스트 24 〉	통과 (1.99ms, 9.64MB)
테스트 25 〉	통과 (0.01ms, 4.17MB)
테스트 26 〉	통과 (0.01ms, 4.17MB)
테스트 27 〉	통과 (0.01ms, 3.59MB)
*/

#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

// 실패율 = 클리어하지 못한 플레이어수 / 스테이지에 도달한 플레이어수
// 실패율이 높은 스테이지로부터 내림차순으로 스테이지 번호

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;

    unordered_map<int, int> m;
    for (auto s : stages) m[s]++;

    vector<pair<float, int>> failure; // {실패율, 스테이지}

    int people = stages.size();
    for (int stage = 1; stage <= N; stage++) {
        if (m.find(stage) == m.end()) failure.push_back({0, stage});
        else {
            int blocked = m[stage];
            failure.push_back({(float)blocked / people, stage});
            people -= blocked;
        }
    }

    auto cmp = [](pair<float, int>& a, pair<float, int>& b) {
        if (a.first == b.first) return a.second < b.second;
        return a.first > b.first;
    };
    sort(failure.begin(), failure.end(), cmp);

    for (auto f : failure) {
        answer.push_back(f.second);
    }

    return answer;
}

